### ISSUE TITLE
Traiding bot Somewhat improved 

### ISSUE THREAD

**Issue by Atakan-24** [Atakan-24/TEST-2#3]
import asyncio
import time
import numpy as np
from talib import RSI, MACD
import logging

# Set up logging
logging.basicConfig(filename='trading.log', level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

# Bitmart API credentials
API_KEY = 'your_api_key'
API_SECRET = 'your_api_secret'

# Function to get latest price of a specific trading pair asynchronously
async def get_latest_price(symbol):
    url = f'https://api-cloud.bitmart.com/spot/v1/ticker?symbol={symbol}'
    async with aiohttp.ClientSession() as session:
        async with session.get(url) as response:
            data = await response.json()
            return float(data['data']['ticker']['last'])

# Function to place a limit buy order with stop-loss and take-profit asynchronously
async def place_limit_buy_order(symbol, quantity, price, stop_loss=None, take_profit=None):
    endpoint = 'https://api-cloud.bitmart.com/spot/v1/submit_order'
    payload = {
        'symbol': symbol,
        'type': '2',  # Limit buy
        'quantity': quantity,
        'price': price,
        'side': 'buy',
        'timeInForce': 'GTC',  # Good 'Til Canceled
        'recvWindow': 5000,
        'timestamp': int(time.time() * 1000)
    }
    if stop_loss:
        payload['stopPrice'] = stop_loss
    if take_profit:
        payload['stopPrice'] = take_profit
    async with aiohttp.ClientSession() as session:
        async with session.post(endpoint, data=payload) as response:
            return await response.json()

# Function to place a limit sell order with stop-loss and take-profit asynchronously
async def place_limit_sell_order(symbol, quantity, price, stop_loss=None, take_profit=None):
    endpoint = 'https://api-cloud.bitmart.com/spot/v1/submit_order'
    payload = {
        'symbol': symbol,
        'type': '2',  # Limit sell
        'quantity': quantity,
        'price': price,
        'side': 'sell',
        'timeInForce': 'GTC',  # Good 'Til Canceled
        'recvWindow': 5000,
        'timestamp': int(time.time() * 1000)
    }
    if stop_loss:
        payload['stopPrice'] = stop_loss
    if take_profit:
        payload['stopPrice'] = take_profit
    async with aiohttp.ClientSession() as session:
        async with session.post(endpoint, data=payload) as response:
            return await response.json()

# Main function to run the trading bot
async def main():
    symbol = 'ushark_usdt'  # Replace with your desired trading pair
    quantity = 10  # Replace with your desired quantity

    while True:
        price = await get_latest_price(symbol)
        if price is not None:
            logging.info(f'Latest price of {symbol}: {price}')

            # Calculate RSI
            rsi_period = 14
            rsi_values = RSI(np.array([price]), timeperiod=rsi_period)
            rsi = rsi_values[0][-1]

            # Calculate MACD
            macd, signal, _ = MACD(np.array([price]), fastperiod=12, slowperiod=26, signalperiod=9)
            macd_diff = macd[-1] - signal[-1]

            logging.info(f'RSI: {rsi}, MACD Difference: {macd_diff}')

            # Advanced risk management: Implementing stop-loss and take-profit orders
            stop_loss = price * 0.95  # 5% stop loss below entry price
            take_profit = price * 1.1  # 10% take profit above entry price

            # Buy if RSI is low and MACD difference is positive
            if rsi <= 30 and macd_diff > 0:
                logging.info('Low RSI and positive MACD difference. Placing buy order...')
                buy_price = price * 0.99  # Buy slightly below current price
                buy_response = await place_limit_buy_order(symbol, quantity, buy_price, stop_loss, take_profit)
                if buy_response:
                    logging.info('Buy order response:', buy_response)

        time.sleep(60)  # Check price every 60 seconds

if __name__ == '__main__':
    asyncio.run(main())
